### PR TITLE
111 in page table of contents

### LINKED ISSUE #111: in-page table of contents
Side nav is optional. Some long-form content will need it, some will not. So the content owner/author will need the ability to add side nav, if they need it—and the body of the page will start at column 3. If not used, start at column 1.

Content Hubs Figma: https://www.figma.com/file/oSetT4LbatRmXlcB2A7V8Z/Content-Hubs-2024?type=design&node-id=891-41409&mode=design&t=FWpXdKc0ZJ2AyKzd-4

Screenshot 2024-02-15 at 17 20 28

Brand Design: https://www.figma.com/file/OFEfvhyXJvYJ6MQZ30699m/v2.0-Brand-Digital-Design-System?type=design&node-id=7625-32246&mode=design&t=K62uHnxv6G3K65XD-0

### PR DESCRIPTION
**Pull request by larsauffarth**
Fix #111

This introduces the side navigation / table of contents for L3 article pages.
As requested, it is a basic version that does not yet track the current position in the page to update the TOC but first focusses on the hash in the url to highlight the active part.

Please provide some feedback on the integration, as it this block renders at the same time as the hero, not subsequently as the remaining page (if that is the desired state?) LHS does not change / improves for desktop compared with the main branch.
Also, where & how do we document blocks that are generated by metadata like this one?

Test URLs:
- Before: https://main--hlx-test--urfuwo.hlx.page/news/0000/the-article
- After: https://111-in-page-table-of-contents--hlx-test--urfuwo.hlx.page/news/0000/the-article

- [x] New Blocks introduced in this PR

Block Name    | Documentation | Library Link
------------- | -------------|----------------
toc | [doc-link](https://sap.sharepoint.com/:w:/r/sites/207899/_layouts/15/Doc.aspx?sourcedoc=%7BF661518B-AC56-470D-8C20-2CB83186CF6A%7D&file=toc.docx&action=default&mobileredirect=true) | [Library Link](https://main--hlx-test--urfuwo.hlx.page/tools/sidekick/library.html?plugin=blocks&path=/tools/sidekick/blocks/toc&index=0)

- [ ] New variations introduced in this PR

### CODE REVIEW

File: blocks/toc/toc.js [hunk 1]
```diff
@@ -0,0 +1,44 @@
+function createEntry(element) {
```
Comment (pahupe) [inline blocks/toc/toc.js:1]: @larsauffarth : Please see if scripts/dom-builder.js might help you here, see social.js as an example

File: templates/article/article.js [hunk 2]
```diff
@@ -12,6 +12,15 @@ function restructureArticle(container, targetClass) {
 function decorate(doc) {
   const main = doc.querySelector('main');
   restructureArticle(main, '.hero');
+  const tocFlag = document.querySelector('meta[name="toc"]');
+  if (tocFlag && tocFlag.content !== 'no' && tocFlag.content !== 'false') {
+    const tocSection = document.createElement('div');
```
Comment (pahupe) [inline templates/article/article.js:17]: Same here: dom-builder.js. But might be a overkill for creating a div with a single CSS class

**Review by pahupe — APPROVED** [review]
I think this is a good first start, and I suggest to do the other changes in a separate "... iteration 2" GH issue.
Two things I think need to be addressed in subsequent tickets:
- What to do with the TOC in small viewports?
- No TOC entry is initially highlighted (not sure if that is desired)

File: templates/article/article.js [hunk 3]
```diff
@@ -12,6 +12,15 @@ function restructureArticle(container, targetClass) {
...
+    const tocBlock = document.createElement('div');
+    tocSection.classList.add('toc-container');
+    tocBlock.classList.add('toc');
+    tocSection.appendChild(tocBlock);
+    main.insertBefore(tocSection, document.querySelector('main > :nth-child(2)'));
```
Comment (saurabh-khare) [inline templates/article/article.js:22]: use `doc` instead of `document` to avoid scope creep

File: blocks/toc/toc.js [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+    block.append(tocElement);
+  }
+}
+
+function setActiveLink() {
+  const links = document.querySelectorAll('.toc li');
+  links.forEach((link) => {
+    const linkHash = link.querySelector('a')?.hash;
+    if (linkHash === window.location.hash) {
+      link.classList.add('active');
```
[the remark this answers is not in the document]
Comment (larsauffarth) [reply blocks/toc/toc.js]: good catch :)

File: blocks/toc/toc.js [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+  return listEl;
+}
+
+export default async function decorate(block) {
+  const mainContent = document.querySelector('main > :nth-child(3)');
+  const headers = mainContent?.querySelectorAll('h2, h3');
+  if (headers.length > 0) {
+    const tocElement = document.createElement('div');
+    const h2 = document.createElement('h2');
+    h2.innerText = 'What\'s on this page';
```
[the remark this answers is not in the document]
Comment (larsauffarth) [reply blocks/toc/toc.js]: I think this is yet to be decided, followed the "cupcake" approach here - should be handled in a second iteration on this component

**Comment by saurabh-khare** [comment]
Screenshot 2024-02-26 at 15 08 35
Don't know if promo headings should also appear as anchor links. But I guess its not for us to decide ;)
